core/arch/arm/kernel/user_ta.c

```diff
@@ -302,6 +303,13 @@ static TEE_Result init_with_ldelf(struct tee_ta_session *sess,
 	utc->dump_entry_func = arg->dump_entry;
 #ifdef CFG_TA_FTRACE_SUPPORT
 	utc->ftrace_entry_func = arg->ftrace_entry;
+	sess->fbuf = arg->fbuf;
+	/*
+	 * Enable accesses to frequency register and physical counter
+	 * register in el0 mode required for timestamping during
+	 * function tracing.
+	 */
+	write_cntkctl(read_cntkctl() | CNTKCTL_PL0PCTEN);
```

Comment:
This is a per-CPU setting, right? So is this the right place?

Reply:
Actually there is only a single instance of this register among secure and non-secure mode. So any configuration during optee_os init might be overridden during Linux boot. That's why I need to configure it during run-time as required.

Reply:
OK, and this single call will enable access for all cores?

Reply:
CNTCKCTL seems to be a per core register. So this call will enable access for a particular core only.

Reply:
So what happens if a different core is scheduled to run ftrace-instrumented code?

Reply:
This register needs to be preserved and restored when switching between worlds. In TF-A `NS_TIMER_SWITCH` controls if the register is handled by the dispatcher or not.

This should also be done in the monitor, it seems we're already doing something related see:
https://github.com/OP-TEE/optee_os/blob/1cdd95a2a46d1573514d04270938de6bd7f52d98/core/arch/arm/sm/sm_a32.S#L43-L46

Reply:
Thanks @jenswi-linaro for the pointer. Following change in TF-A should allow optee to control access to timer registers independently of Linux:

```
diff --git a/services/spd/opteed/opteed.mk b/services/spd/opteed/opteed.mk
index b1b2175..e0d3e03 100644
--- a/services/spd/opteed/opteed.mk
+++ b/services/spd/opteed/opteed.mk
@@ -13,3 +13,6 @@ SPD_SOURCES           :=      services/spd/opteed/opteed_common.c     \
                                services/spd/opteed/opteed_pm.c
 
 NEED_BL32              :=      yes
+
+# required so that optee code can control access to the timer registers
+NS_TIMER_SWITCH                :=      1
```

Also, we could configure `CNTCKCTL` during optee_os init for every core and that should persist during runtime.

Reply:
TF-A patch here: https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/1711

Reply:
@b49020 thanks for submitting the TF-A patch, problem solved for Aarch64.

As for Aarch32, it seems that a couple of instance of `#ifdef CFG_SM_NO_CYCLE_COUNTING` need to be replaced with `#if defined(CFG_SM_NO_CYCLE_COUNTING) || defined(CFG_TA_FTRACE_SUPORT)`?

Reply:
ok.

Reply:
Sorry, I did recheck for this. It looks like `CNTKCTL` is a banked register (separate secure and non-secure copy) in case for ARMv7-A (see arm_architecture_reference_manual, section: B8.1.5 Timers). So we don't need to save/restore generic timer registers.

Reply:
I'm confused. B4.1.26 (CNTKCTL, Timer PL1 Control register, VMSA) says: "If the implementation includes the Security Extensions, this register is Common.". Compare with B4.1.29 (CNTP_TVAL, PL1 Physical TimerValue register, VMSA) which says "If the implementation includes the Security Extensions, this register is Banked."

So to me it sounds like the timer value is banked but the control register is not.

Reply:
Yes you are correct. Spec seems to be confusing. Will add corresponding implementation in secure monitor for ARMv7.

Reply:
Done.